sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/parsers.scala

```diff
@@ -105,11 +109,61 @@ abstract class AbstractParser extends DataTypeParserInterface with Logging {
           messageParameters = e.getMessageParameters.asScala.toMap,
           queryContext = e.getQueryContext)
     }
+    finally {
+      if (conf.releaseAntlrCacheAfterParsing) {
+        AbstractSqlParser.refreshParserCaches()
+      }
+    }
   }
 
   private def conf: SqlApiConf = SqlApiConf.get
 }
 
+object AbstractSqlParser {
+  private case class AntlrCaches(atn: ATN) {
+    val predictionContextCache: PredictionContextCache = new PredictionContextCache
+    val decisionToDFA: Array[DFA] = AntlrCaches.makeDecisionToDFA(atn)
+    def installCaches(parser: SqlBaseParser): Unit = {
+      parser.setInterpreter(
+        new ParserATNSimulator(parser, atn, decisionToDFA, predictionContextCache))
+    }
+  }
+
+  private object AntlrCaches {
+    private def makeDecisionToDFA(atn: ATN): Array[DFA] = {
+      val decisionToDFA = new Array[DFA](atn.getNumberOfDecisions)
+      for (i <- 0 until atn.getNumberOfDecisions) {
+        decisionToDFA(i) = new DFA(atn.getDecisionState(i), i)
+      }
+      decisionToDFA
+    }
+  }
+
+  private val parserCaches = new AtomicReference[AntlrCaches](AntlrCaches(SqlBaseParser._ATN))
+
+  /**
+   * Install the parser caches into the given parser.
+   *
+   * This method should be called before parsing any input.
+   */
+  def installCaches(parser: SqlBaseParser): Unit = parserCaches.get().installCaches(parser)
+
+  /**
+   * Drop the existing parser caches and create a new one.
+   *
+   * ANTLR retains caches in its parser that are never released.  This speeds up parsing of future
+   * input, but it can consume a lot of memory depending on the input seen so far.
```

Comment:
Does this cache work within a query or only for future queries?

Reply:
It's my understanding that it works for future tokens in the same query as well as for future queries.

`ParserInterpreter.parse` calls `ParserInterpreter.visitState`, which calls `ParserInterpreter.visitDecisionState`, which calls `ParserATNSimulator.adaptivePredict`.  The `ParserATNSimulator` was constructed with a reference to the static DFA cache and `ParserATNSimulator.adaptivePredict` reads/updates that cache via `ParserATNSimulator.addDFAState` which calls `ATNConfigSet.optimizeConfigs`, which calls `ATNSimulator.getCachedContext`, which calls `PredictionContext.getCachedContext`, which calls `PredictionContextCache.add` and `PredictionContextCache.get`.

https://github.com/antlr/antlr4/blob/69cfd8e49b911c4a0483035b100fd438f1a55841/runtime/Java/src/org/antlr/v4/runtime/atn/PredictionContext.java#L544-L605

So, in short, any time there's a prediction during parsing, whether it's later in an input stream or in a subsequent input stream, it's using this cache.

Reply:
When I say:

> The `ParserATNSimulator` was constructed with a reference to the static DFA cache

This is best seen in the ANTLR generated code.  Here's a snippet:

```java
public class SqlBaseParser extends Parser {
        static { RuntimeMetaData.checkVersion("4.8", RuntimeMetaData.VERSION); }

        protected static final DFA[] _decisionToDFA;
        protected static final PredictionContextCache _sharedContextCache =
                new PredictionContextCache();
...
        public SqlBaseParser(TokenStream input) {
                super(input);
                _interp = new ParserATNSimulator(this,_ATN,_decisionToDFA,_sharedContextCache);
        }
...
        static {
                _decisionToDFA = new DFA[_ATN.getNumberOfDecisions()];
                for (int i = 0; i < _ATN.getNumberOfDecisions(); i++) {
                        _decisionToDFA[i] = new DFA(_ATN.getDecisionState(i), i);
                }
        }
}
```